ddl/column.go

```diff
@@ -93,57 +99,116 @@ func (d *ddl) createColumnInfo(tblInfo *model.TableInfo, colInfo *model.ColumnIn
+}
+
+func checkHasDuplicateColumnName(columnsInfo []*model.ColumnInfo) error {
+	columnNames := map[string]bool{}
+	for i := 0; i < len(columnsInfo); i++ {
+		name := columnsInfo[i].Name.L
+		if columnNames[name] {
+			return infoschema.ErrColumnExists.GenByArgs(name)
+		}
+		columnNames[name] = true
+	}
+	return nil
+}
+
+// getColumnInfos get all columnInfos correspond to column
+func (d *ddl) getColumnInfos(columns []*model.ColumnInfo, positions []*ast.ColumnPosition, job *model.Job, tblInfo *model.TableInfo,
+	lastColOffset int) ([]*model.ColumnInfo, error) {
+	columnInfos := []*model.ColumnInfo{}
+	var err error
+	for idx, col := range columns {
+		colInfo, _ := findCol(tblInfo.Columns, col.Name.L)
+		pos := &ast.ColumnPosition{Tp: ast.ColumnPositionNone}
+		if idx < len(positions) {
+			pos = positions[idx]
+		}
+
+		if colInfo != nil {
+			if colInfo.State == model.StatePublic {
+				// We already have a column with the same column name.
+				job.State = model.JobCancelled
+				return nil, infoschema.ErrColumnExists.GenByArgs(col.Name)
+			}
+		} else {
+			// mark sequentially the offset of new column as the last column added idx
+			colInfo, _, err = d.createColumnInfo(tblInfo, col, pos, lastColOffset+idx)
+			if err != nil {
+				job.State = model.JobCancelled
+				return nil, errors.Trace(err)
+			}
+		}
+		columnInfos = append(columnInfos, colInfo)
+	}
+	return columnInfos, nil
+}
+
+func (d *ddl) onAddColumns(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	schemaID := job.SchemaID
 	tblInfo, err := getTableInfo(t, job, schemaID)
 	if err != nil {
 		return ver, errors.Trace(err)
 	}
 
-	col := &model.ColumnInfo{}
-	pos := &ast.ColumnPosition{}
-	offset := 0
-	err = job.DecodeArgs(col, pos, &offset)
-	if err != nil {
+	columns := []*model.ColumnInfo{}
+	positions := []*ast.ColumnPosition{}
+	lastColOffset := 0
+	if err = job.DecodeArgs(&columns, &positions, &lastColOffset); err != nil {
 		job.State = model.JobCancelled
 		return ver, errors.Trace(err)
 	}
+	// First run of this job, get the offset of last column.
+	if lastColOffset == 0 {
+		lastColOffset = len(tblInfo.Columns)
+	}
 
-	columnInfo := findCol(tblInfo.Columns, col.Name.L)
-	if columnInfo != nil {
-		if columnInfo.State == model.StatePublic {
-			// We already have a column with the same column name.
-			job.State = model.JobCancelled
-			return ver, infoschema.ErrColumnExists.GenByArgs(col.Name)
-		}
-	} else {
-		columnInfo, offset, err = d.createColumnInfo(tblInfo, col, pos)
-		if err != nil {
-			job.State = model.JobCancelled
-			return ver, errors.Trace(err)
-		}
-		// Set offset arg to job.
-		if offset != 0 {
-			job.Args = []interface{}{columnInfo, pos, offset}
-		}
+	if err = checkHasDuplicateColumnName(columns); err != nil {
+		job.State = model.JobCancelled
+		return ver, errors.Trace(err)
 	}
 
-	originalState := columnInfo.State
-	switch columnInfo.State {
+	columnsInfo, err := d.getColumnInfos(columns, positions, job, tblInfo, lastColOffset)
```

Comment:
The new added columns is hidden by setting its offset is greater than the offset of last column
of the table. So it should save the lastColOffset before the createColumnInfo .